### PR TITLE
refactor DPMSolverMultistepScheduler using sigmas

### PR DESCRIPTION
**Pull request by yiyixuxu**
I think I overcomplicated things with this https://github.com/huggingface/diffusers/pull/4690 by trying to follow k-diffusion implementation. 

trying it again with a simpler approach here: we simply calculate `lambda_t`, `sigma_t` and `alpha_t` from `sigma`
### to-do
- [x] dpmsolver_multistep
- [x] dpmsolver_singlestep
- [x] DEIS 
- [x] UniPC

### testing
__Notes:__
I compared the results against the current implementation (for dpm multistep only). There is some slight numerical difference when using k-sigmas; for example, in the last testing examples below on the 4th row (`sde-dpmsolver++, use_karras_sigma=True`), you can visually see the outputs are slightly different. 
Also, I had to loosen a test here https://github.com/huggingface/diffusers/pull/4986#discussion_r1324034732. 
However, the new implementation is more accurate by using the sigma directly. 

```python
import torch
from diffusers import StableDiffusionKDiffusionPipeline, DPMSolverMultistepScheduler, StableDiffusionPipeline
import gc

# make sure to update your branch name 
branch = 'dpm-mstep-sigma-2'
#branch = "main"


# test 1: dpmsolver_+++, use_karras_sigma=False 
seed = 33

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
pipe.scheduler = DPMSolverMultistepScheduler.from_config(
    pipe.scheduler.config, use_karras_sigmas=False
)

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]

image.save(f"[{branch}]_test_dpmsolver++.png")


# test 2: dpmsolver_+++, use_karras_sigma=True 
seed = 33

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
pipe.scheduler = DPMSolverMultistepScheduler.from_config(
    pipe.scheduler.config, use_karras_sigmas=True
)

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]

image.save(f"[{branch}]_test_dpmsolver++_k_sigma.png")

# # test 3: sde-dpmsolver++, use_karras_sigma=False 

seed = 33

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
pipe.scheduler = DPMSolverMultistepScheduler.from_config(
    pipe.scheduler.config, use_karras_sigmas=False, algorithm_type="sde-dpmsolver++"
)

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]

image.save(f"[{branch}]_test_sde-dpmsolver++.png")

## test 4: sde-dpmsolver++, use_karras_sigma=True
seed = 33

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
pipe.scheduler = DPMSolverMultistepScheduler.from_config(
    pipe.scheduler.config, use_karras_sigmas=True, algorithm_type="sde-dpmsolver++"
)

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]
image.save(f"[{branch}]_test_sde-dpmsolver++_k_sigma.png")
```
main            |  this PR
:-------------------------:|:-------------------------:
dpm-testing _test_dpmsolver++ |  dpm-mstep-sigma-2 _test_dpmsolver++
dpm-testing _test_dpmsolver++_k_sigma |  dpm-mstep-sigma-2 _test_dpmsolver++_k_sigma
dpm-testing _test_sde-dpmsolver++ | dpm-mstep-sigma-2 _test_sde-dpmsolver++
dpm-testing _test_sde-dpmsolver++_k_sigma |  dpm-mstep-sigma-2 _test_sde-dpmsolver++_k_sigma

### CODE REVIEW

File: tests/schedulers/test_scheduler_dpm_multi.py [hunk 1]
```diff
@@ -227,7 +229,7 @@ def test_full_loop_with_karras_and_v_prediction(self):
         sample = self.full_loop(prediction_type="v_prediction", use_karras_sigmas=True)
         result_mean = torch.mean(torch.abs(sample))
 
-        assert abs(result_mean.item() - 0.2096) < 1e-3
+        assert abs(result_mean.item() - 0.2096) < 2e-3
```
Comment (yiyixuxu) [inline tests/schedulers/test_scheduler_dpm_multi.py:232]: would this be ok? I looked into it, think there is no difference in calculation

File: tests/schedulers/test_scheduler_dpm_multi_inverse.py [hunk 2]
```diff
@@ -54,6 +54,7 @@ def check_over_configs(self, time_step=0, **config):
 
             output, new_output = sample, sample
             for t in range(time_step, time_step + scheduler.config.solver_order + 1):
+                t = scheduler.timesteps[t]
```
Comment (yiyixuxu) [inline tests/schedulers/test_scheduler_dpm_multi_inverse.py:57]: We used `0`,`1`,`2` as `timestep` index here, but they are not even in the `self.timestepes`.  In the previous implementation, it will default to the last timesteps when it is outside of the timesteps range, but I don't think it is intended. I changed it here, I think it makes more sense this way. Let me know if it's not the case

File: tests/schedulers/test_scheduler_dpm_multi_inverse.py [hunk 3]
```diff
@@ -222,7 +223,7 @@ def test_full_loop_with_karras_and_v_prediction(self):
         sample = self.full_loop(prediction_type="v_prediction", use_karras_sigmas=True)
         result_mean = torch.mean(torch.abs(sample))
 
-        assert abs(result_mean.item() - 1.7833) < 1e-3
```
Comment (yiyixuxu) [inline tests/schedulers/test_scheduler_dpm_multi_inverse.py:225]: Would this be ok?

**Comment by yiyixuxu** [comment]
@patrickvonplaten can take a review here if you have time!

**Comment by yiyixuxu** [comment]
this PR also fix https://github.com/huggingface/diffusers/issues/4122

this script now return `100`

```python
from diffusers import StableDiffusionXLPipeline
from diffusers.schedulers import DPMSolverMultistepScheduler
import torch

scheduler = DPMSolverMultistepScheduler(
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    prediction_type="epsilon",
    num_train_timesteps=1000,
    trained_betas=None,
    thresholding=False,
    algorithm_type="dpmsolver++",
    solver_type="midpoint",
    lower_order_final=True,
    use_karras_sigmas=True,
)


pipe = StableDiffusionXLPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", scheduler=scheduler, variant="fp16", use_safetensors=True,
    torch_dtype=torch.float16,
)
pipe.enable_model_cpu_offload()


random_generator = torch.Generator()
random_generator.manual_seed(12345)
prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"
image = pipe(prompt=prompt, output_type="pil",
    num_inference_steps=100, generator=random_generator).images[0]
print(len(pipe.timesteps))
```

File: src/diffusers/schedulers/scheduling_dpmsolver_multistep.py [hunk 4]
```diff
@@ -337,9 +355,7 @@ def _convert_to_karras(self, in_sigmas: torch.FloatTensor, num_inference_steps)
         sigmas = (max_inv_rho + ramp * (min_inv_rho - max_inv_rho)) ** rho
         return sigmas
 
-    def convert_model_output(
-        self, model_output: torch.FloatTensor, timestep: int, sample: torch.FloatTensor
-    ) -> torch.FloatTensor:
+    def convert_model_output(self, model_output: torch.FloatTensor, sample: torch.FloatTensor) -> torch.FloatTensor:
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:358]:
```suggestion
    def convert_model_output(self, model_output: torch.FloatTensor, *args, sample: torch.FloatTensor = None, **kwargs) -> torch.FloatTensor:
    timestep = args[0] if len(args) > 0 else kwargs.pop("timestep", None)
    if timestep is not None:
        deprecate("Passing `timesteps` is deprecated and has no effect as model output conversion is now handled via an internal counter `self.step_index`")
```

File: src/diffusers/schedulers/scheduling_dpmsolver_multistep.py [hunk 5]
```diff
@@ -469,8 +490,6 @@ def dpm_solver_first_order_update(
     def multistep_dpm_solver_second_order_update(
         self,
         model_output_list: List[torch.FloatTensor],
-        timestep_list: List[int],
-        prev_timestep: int,
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:473]: add *args and **kwargs to deprecate

File: src/diffusers/schedulers/scheduling_dpmsolver_multistep.py [hunk 6]
```diff
@@ -564,8 +595,6 @@ def multistep_dpm_solver_second_order_update(
     def multistep_dpm_solver_third_order_update(
         self,
         model_output_list: List[torch.FloatTensor],
-        timestep_list: List[int],
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:567]: add *args and **kwargs to deprecate

**Review by patrickvonplaten — COMMENTED** [review]
Very nice! The approach looks great, just left some comments regarding warning the user a bit better that some public methods have a changed signature.

Could you also apply these changes to `dpmsolver_singlestep`, DEIS and UniPC?

File: src/diffusers/schedulers/scheduling_dpmsolver_multistep.py [hunk 7]
```diff
@@ -242,19 +251,19 @@ def set_timesteps(self, num_inference_steps: int = None, device: Union[str, torc
             )
 
         sigmas = np.array(((1 - self.alphas_cumprod) / self.alphas_cumprod) ** 0.5)
+        log_sigmas = np.log(sigmas)
+
         if self.config.use_karras_sigmas:
-            log_sigmas = np.log(sigmas)
+            sigmas = np.flip(sigmas).copy()
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:257]: reversed order of `sigmas` here
fix https://github.com/huggingface/diffusers/issues/5077#issuecomment-1723931694

File: src/diffusers/schedulers/scheduling_dpmsolver_multistep.py [hunk 8]
```diff
@@ -323,6 +335,12 @@ def _sigma_to_t(self, sigma, log_sigmas):
         t = t.reshape(sigma.shape)
         return t
 
+    def _sigma_to_alpha_sigma_t(self, sigma):
```
[the remark this answers is not in the document]
Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:338]: actually, originated here and not copied from anywhere 😛

File: tests/schedulers/test_scheduler_unipc.py [hunk 9]
```diff
@@ -241,11 +242,3 @@ def test_fp16_support(self):
             sample = scheduler.step(residual, t, sample).prev_sample
 
         assert sample.dtype == torch.float16
-
-    def test_unique_timesteps(self, **config):
```
Comment (yiyixuxu) [inline tests/schedulers/test_scheduler_unipc.py:245]: don't need this test anymore because we allow duplicated timesteps now

**Comment by yiyixuxu** [comment]
@patrickvonplaten another review please :)

File: tests/schedulers/test_scheduler_dpm_multi.py [hunk 10]
```diff
@@ -264,10 +266,10 @@ def test_fp16_support(self):
...
 
-    def test_unique_timesteps(self, **config):
+    def test_duplicated_timesteps(self, **config):
         for scheduler_class in self.scheduler_classes:
             scheduler_config = self.get_scheduler_config(**config)
             scheduler = scheduler_class(**scheduler_config)
 
             scheduler.set_timesteps(scheduler.config.num_train_timesteps)
-            assert len(scheduler.timesteps.unique()) == scheduler.num_inference_steps
+            assert len(scheduler.timesteps) == scheduler.num_inference_steps
```
Comment (patrickvonplaten) [inline tests/schedulers/test_scheduler_dpm_multi.py:275]: nice!

**Review by patrickvonplaten — APPROVED** [review]
Very nice!

**Comment by rvorias** [comment]
@yiyixuxu This refactor has broken img2img for `unipc`

Using unipc code before this commit:
image

Using unipc after this commit:
image

**Comment by ljk1291** [comment]
> @yiyixuxu This refactor has broken stable diffusion (1.5) img2img for `unipc`
> 
> both use denoising strength = 0.6
> 
> Using unipc code before this commit: image
> 
> Using unipc after this commit: image

Same with DPMSolverMultistepScheduler using Karras sigmas. Just returns noise with anything under 1.0 denoising strength

**Comment by yiyixuxu** [comment]
@ljk1291
oh, thanks for reporting! I'm looking into it now!
if you have a reproducible example and can open a new issue - that would be great!

File: src/diffusers/schedulers/scheduling_dpmsolver_multistep.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (krahnikblis) [reply src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:338]: hi @yiyixuxu, question from a casual observer here, following along and converting UniPC into jax... it seems the `_sigma_to_alpha_sigma_t` is used up to 9X in every scheduler step (once in `convert_model_output`, twice in s0 and t versions in each the C and P multistep update functions, and up to twice again in each according to the `order` determination of that step, which would most commonly be 2 or 3, so 1X or 2X more again)

is this function used elsewhere in other schedulers in a way that makes it impossible to run the calculation once and set all `sigmas` and `alphas` within the `set_timesteps` function? from what i can tell during my tinkering, `alphas`, `lambdas`, `sigmas` (and scheduler-specific things like `R`, `rks`, `rhos`, `order`, etc.) are only dependent on `timestep` and input parameters, not any data/sample values during the `step` process, so they can be created in advance and then referenced from the `state` (or `self` for torch) variable using `[step_index]` instead of `[timestep]`, which all could maybe save some compute (and lines of code?)

**Comment by burgalon** [comment]
it seems like this bug still resurfaces somehow when using the legacy inpainting pipeline for some reason

**Comment by yiyixuxu** [comment]
Hi @burgalon 

can you be more specific? could you maybe open a bug report and provide an example?

thanks!

YiYi

**Comment by burgalon** [comment]
oops... happened to post here https://github.com/huggingface/diffusers/issues/4631#issuecomment-1824783221

File: src/diffusers/schedulers/scheduling_dpmsolver_multistep.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:338]: hey! sorry I missed this comment. a good suggestion! feel free to open an PR, otherwise I will find a time to work on this soon
